File: ruby/lib/callisto/resources/whatsapp.rb
# frozen_string_literal: true

module Callisto
  # WhatsApp resource.
  class WhatsAppResource
    # @param transport [Transport]
    def initialize(transport)
      @t = transport
    end

    # Creates a WhatsApp instance.
    #
    # @param name [String] instance display name
    # @param phone_number [String, nil]
    # @param webhook_url [String, nil]
    # @param idempotency_key [String, nil]
    # @return [WhatsAppInstance]
    def create_instance(name:, phone_number: nil, webhook_url: nil, idempotency_key: nil)
      body = drop_nil({
        "name" => name, "phone_number" => phone_number,
        "webhook_url" => webhook_url, "idempotency_key" => idempotency_key
      })
      WhatsAppInstance.from_hash(@t.request("POST", "/whatsapp/instances", body: body))
    end

    # Lists WhatsApp instances.
    #
    # @param page [Integer]
    # @return [Paginated] of WhatsAppInstance
    def list_instances(page: 1)
      data = @t.request("GET", "/whatsapp/instances", query: { "page" => page })
      Paginated.from_hash(data, WhatsAppInstance.method(:from_hash))
    end

    # Fetches a single instance.
    #
    # @param code [String]
    # @return [WhatsAppInstance]
    def get_instance(code)
      WhatsAppInstance.from_hash(@t.request("GET", "/whatsapp/#{code}"))
    end

    # Fetches the QR code used to link the instance. Returns the raw decoded payload.
    #
    # @param code [String]
    # @return [Object] raw payload
    def get_qr(code)
      @t.request("GET", "/whatsapp/#{code}/qr")
    end

    # Fetches the connection status of an instance. Returns the raw decoded payload.
    #
    # @param code [String]
    # @return [Object] raw payload
    def get_status(code)
      @t.request("GET", "/whatsapp/#{code}/status")
    end

    # Lists messages for an instance.
    #
    # @param code [String]
    # @param started_at [String, nil]
    # @param ended_at [String, nil]
    # @param page [Integer, nil]
    # @param per_page [Integer, nil]
    # @return [Paginated] of WhatsAppMessage
    def list_messages(code, started_at: nil, ended_at: nil, page: nil, per_page: nil)
      data = @t.request("GET", "/whatsapp/#{code}/messages", query: {
        "started_at" => started_at, "ended_at" => ended_at,
        "page" => page, "per_page" => per_page
      })
      Paginated.from_hash(data, WhatsAppMessage.method(:from_hash))
    end

    # Fetches a single WhatsApp message.
    #
    # @param message_id [String]
    # @return [WhatsAppMessage]
    def get_message(message_id)
      WhatsAppMessage.from_hash(@t.request("GET", "/whatsapp/messages/#{message_id}"))
    end

    # Sends a text message.
    #
    # @return [SendWaResult]
    def send_text(code, to:, message:, scheduled_at: nil)
      body = drop_nil({ "to" => to, "message" => message, "scheduled_at" => scheduled_at })
      SendWaResult.from_hash(@t.request("POST", "/whatsapp/#{code}/send/text", body: body))
    end

    # Sends a media message.
    #
    # @param type [String] WhatsAppMediaType value or raw string
    # @return [SendWaResult]
    def send_media(code, to:, type:, media_url:, caption: nil, filename: nil, scheduled_at: nil)
      body = drop_nil({
        "to" => to, "type" => type, "media_url" => media_url,
        "caption" => caption, "filename" => filename, "scheduled_at" => scheduled_at
      })
      SendWaResult.from_hash(@t.request("POST", "/whatsapp/#{code}/send/media", body: body))
    end

    # Sends an interactive buttons message.
    #
    # @return [SendWaResult]
    def send_buttons(code, to:, body:, buttons:, header: nil, footer: nil, scheduled_at: nil)
      payload = drop_nil({
        "to" => to, "body" => body, "buttons" => buttons,
        "header" => header, "footer" => footer, "scheduled_at" => scheduled_at
      })
      SendWaResult.from_hash(@t.request("POST", "/whatsapp/#{code}/send/buttons", body: payload))
    end

    # Sends a location message.
    #
    # @return [SendWaResult]
    def send_location(code, to:, latitude:, longitude:, name: nil, address: nil, scheduled_at: nil)
      payload = drop_nil({
        "to" => to, "latitude" => latitude, "longitude" => longitude,
        "name" => name, "address" => address, "scheduled_at" => scheduled_at
      })
      SendWaResult.from_hash(@t.request("POST", "/whatsapp/#{code}/send/location", body: payload))
    end

    # Sends an interactive list message.
    #
    # @return [SendWaResult]
    def send_list(code, to:, body:, button_text:, sections:, header: nil, footer: nil, scheduled_at: nil)
      payload = drop_nil({
        "to" => to, "body" => body, "button_text" => button_text, "sections" => sections,
        "header" => header, "footer" => footer, "scheduled_at" => scheduled_at
      })
      SendWaResult.from_hash(@t.request("POST", "/whatsapp/#{code}/send/list", body: payload))
    end

    private

    def drop_nil(hash)
      hash.reject { |_, v| v.nil? }
    end
  end
end

File: java/src/main/java/com/callisto/sdk/resources/WhatsApp.java
package com.callisto.sdk.resources;

import com.callisto.sdk.http.Transport;
import com.callisto.sdk.models.Paginated;
import com.callisto.sdk.models.SendWaResult;
import com.callisto.sdk.models.WhatsAppInstance;
import com.callisto.sdk.models.WhatsAppMessage;
import com.fasterxml.jackson.databind.JsonNode;

import java.util.LinkedHashMap;
import java.util.Map;

/** The WhatsApp resource. */
public class WhatsApp {

    private final Transport transport;

    public WhatsApp(Transport transport) {
        this.transport = transport;
    }

    /** Creates a WhatsApp instance with only a display name. */
    public WhatsAppInstance createInstance(String name) {
        return createInstance(name, null, null, null);
    }

    /**
     * Creates a WhatsApp instance.
     *
     * @param name           instance display name.
     * @param phoneNumber    optional phone number to attach.
     * @param webhookUrl     optional webhook URL for incoming events.
     * @param idempotencyKey optional key to safely retry creation.
     */
    public WhatsAppInstance createInstance(String name, String phoneNumber,
                                           String webhookUrl, String idempotencyKey) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("name", name);
        putIfPresent(body, "phone_number", phoneNumber);
        putIfPresent(body, "webhook_url", webhookUrl);
        putIfPresent(body, "idempotency_key", idempotencyKey);
        return transport.mapper().convertValue(
                transport.request("POST", "/whatsapp/instances", body, null), WhatsAppInstance.class);
    }

    /** Lists WhatsApp instances starting at page 1. */
    public Paginated<WhatsAppInstance> listInstances() {
        return listInstances(1);
    }

    /**
     * Lists WhatsApp instances.
     *
     * @param page page number (defaults to {@code 1}).
     */
    public Paginated<WhatsAppInstance> listInstances(int page) {
        Map<String, Object> query = new LinkedHashMap<>();
        query.put("page", page);
        return Paginated.fromJson(
                transport.request("GET", "/whatsapp/instances", null, query),
                transport.mapper(), WhatsAppInstance.class);
    }

    /**
     * Fetches a single instance.
     *
     * @param code instance code.
     */
    public WhatsAppInstance getInstance(String code) {
        return transport.mapper().convertValue(
                transport.request("GET", "/whatsapp/" + code, null, null), WhatsAppInstance.class);
    }

    /**
     * Fetches the QR code used to link the instance. Returns the raw {@link JsonNode}.
     *
     * @param code instance code.
     */
    public JsonNode getQr(String code) {
        return transport.request("GET", "/whatsapp/" + code + "/qr", null, null);
    }

    /**
     * Fetches the connection status of an instance. Returns the raw {@link JsonNode}.
     *
     * @param code instance code.
     */
    public JsonNode getStatus(String code) {
        return transport.request("GET", "/whatsapp/" + code + "/status", null, null);
    }

    /** Lists messages for an instance with no filters. */
    public Paginated<WhatsAppMessage> listMessages(String code) {
        return listMessages(code, null, null, null, null);
    }

    /**
     * Lists messages for an instance.
     *
     * @param code      instance code.
     * @param startedAt optional start date/time filter.
     * @param endedAt   optional end date/time filter.
     * @param page      optional page number.
     * @param perPage   optional items per page.
     */
    public Paginated<WhatsAppMessage> listMessages(String code, String startedAt, String endedAt,
                                                   Integer page, Integer perPage) {
        Map<String, Object> query = new LinkedHashMap<>();
        query.put("started_at", startedAt);
        query.put("ended_at", endedAt);
        query.put("page", page);
        query.put("per_page", perPage);
        return Paginated.fromJson(
                transport.request("GET", "/whatsapp/" + code + "/messages", null, query),
                transport.mapper(), WhatsAppMessage.class);
    }

    /**
     * Fetches a single WhatsApp message.
     *
     * @param messageId the message ID.
     */
    public WhatsAppMessage getMessage(String messageId) {
        return transport.mapper().convertValue(
                transport.request("GET", "/whatsapp/messages/" + messageId, null, null),
                WhatsAppMessage.class);
    }

    /** Sends a text message. */
    public SendWaResult sendText(WaTextRequest request) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("to", request.getTo());
        body.put("message", request.getMessage());
        putIfPresent(body, "scheduled_at", request.getScheduledAt());
        return post(request.getCode(), "text", body);
    }

    /** Sends a media message. */
    public SendWaResult sendMedia(WaMediaRequest request) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("to", request.getTo());
        body.put("type", request.getType());
        body.put("media_url", request.getMediaUrl());
        putIfPresent(body, "caption", request.getCaption());
        putIfPresent(body, "filename", request.getFilename());
        putIfPresent(body, "scheduled_at", request.getScheduledAt());
        return post(request.getCode(), "media", body);
    }

    /** Sends an interactive buttons message. */
    public SendWaResult sendButtons(WaButtonsRequest request) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("to", request.getTo());
        body.put("body", request.getBody());
        body.put("buttons", request.getButtons());
        putIfPresent(body, "header", request.getHeader());
        putIfPresent(body, "footer", request.getFooter());
        putIfPresent(body, "scheduled_at", request.getScheduledAt());
        return post(request.getCode(), "buttons", body);
    }

    /** Sends a location message. */
    public SendWaResult sendLocation(WaLocationRequest request) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("to", request.getTo());
        body.put("latitude", request.getLatitude());
        body.put("longitude", request.getLongitude());
        putIfPresent(body, "name", request.getName());
        putIfPresent(body, "address", request.getAddress());
        putIfPresent(body, "scheduled_at", request.getScheduledAt());
        return post(request.getCode(), "location", body);
    }

    /** Sends an interactive list message. */
    public SendWaResult sendList(WaListRequest request) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("to", request.getTo());
        body.put("body", request.getBody());
        body.put("button_text", request.getButtonText());
        body.put("sections", request.getSections());
        putIfPresent(body, "header", request.getHeader());
        putIfPresent(body, "footer", request.getFooter());
        putIfPresent(body, "scheduled_at", request.getScheduledAt());
        return post(request.getCode(), "list", body);
    }

    private SendWaResult post(String code, String kind, Map<String, Object> body) {
        return transport.mapper().convertValue(
                transport.request("POST", "/whatsapp/" + code + "/send/" + kind, body, null),
                SendWaResult.class);
    }

    private static void putIfPresent(Map<String, Object> body, String key, Object value) {
        if (value != null) {
            body.put(key, value);
        }
    }
}
